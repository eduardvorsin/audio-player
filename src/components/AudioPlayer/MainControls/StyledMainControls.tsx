import styled from "styled-components";

interface StyledMainControlsProps {
	className?: string,
	nextAndPrevious: boolean,
}

export const StyledMainControls = styled.div<StyledMainControlsProps>`
	align-self:center;
	display: flex;
	justify-content: ${({ nextAndPrevious }) => nextAndPrevious ? 'center' : 'flex-start'};
	
	& > *:not(:last-child){
		margin-right: 10px;
	}
`;
