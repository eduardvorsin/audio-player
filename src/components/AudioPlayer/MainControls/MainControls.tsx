import React from "react";
import { StyledMainControls } from "./StyledMainControls";
import { Button } from "../../UI/Button/Button";

import { ReactComponent as PlayIcon } from '../../../assets/images/icons/play.svg';
import { ReactComponent as PauseIcon } from '../../../assets/images/icons/pause.svg';
import { ReactComponent as PrevIcon } from '../../../assets/images/icons/prev.svg';
import { ReactComponent as NextIcon } from '../../../assets/images/icons/next.svg';

interface MainControlsProps {
	isPlayed: boolean,
	showNextAndPreviousControls?: boolean,
	togglePlaying?: React.MouseEventHandler<HTMLButtonElement>,
	onClickPrevious?: React.MouseEventHandler<HTMLButtonElement>,
	onClickNext?: React.MouseEventHandler<HTMLButtonElement>,
}

export const MainControls = React.memo<MainControlsProps>(({
	isPlayed,
	showNextAndPreviousControls = false,
	togglePlaying = () => { },
	onClickPrevious = () => { },
	onClickNext = () => { },
}) => {
	const playButtonIcon = isPlayed ? <PauseIcon /> : <PlayIcon />;

	const previousClickHandler: React.MouseEventHandler<HTMLButtonElement> = (e) => {
		onClickPrevious(e);
	}

	const nextClickHandler: React.MouseEventHandler<HTMLButtonElement> = (e) => {
		onClickNext(e);
	}

	return (
		<StyledMainControls
			nextAndPrevious={showNextAndPreviousControls}
			data-testid='main-controls'
		>
			{showNextAndPreviousControls &&
				<Button
					primary
					withoutVisibleText
					startIcon={<PrevIcon />}
					onClick={previousClickHandler}
				>
					prev track
				</Button>
			}
			<Button
				primary
				withoutVisibleText
				startIcon={playButtonIcon}
				onClick={togglePlaying}
			>
				{isPlayed ? 'pause' : 'play'}
			</Button>
			{showNextAndPreviousControls &&
				<Button
					primary
					withoutVisibleText
					startIcon={<NextIcon />}
					onClick={nextClickHandler}
				>
					next track
				</Button>
			}
		</StyledMainControls>
	)
});
