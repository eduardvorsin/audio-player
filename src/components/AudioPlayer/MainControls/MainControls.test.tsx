import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { MainControls } from "./MainControls";

describe('Volume Controls component tests', () => {
  test('renders correctly', () => {
    const mockFn = jest.fn<void, [React.MouseEvent<HTMLButtonElement>]>();
    render(
      <MainControls
        togglePlaying={mockFn}
        onClickPrevious={mockFn}
        onClickNext={mockFn}
        showNextAndPreviousControls
        isPlayed
      />
    );

    expect(screen.getAllByRole<HTMLButtonElement>('button')).toHaveLength(3);
  });

  test('renders with isPlayed prop', () => {
    render(
      <MainControls
        isPlayed
      />
    );

    expect(screen.getAllByRole<HTMLButtonElement>('button')).toHaveLength(1);
  });

  test('when the play button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [React.MouseEvent<HTMLButtonElement>]>();
    render(
      <MainControls
        isPlayed
        togglePlaying={mockFn}
      />
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button'));

    expect(mockFn).toBeCalledTimes(1);
  });

  test('when the prev button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [React.MouseEvent<HTMLButtonElement>]>();
    render(
      <MainControls
        isPlayed
        onClickPrevious={mockFn}
        showNextAndPreviousControls
      />
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button', { name: 'prev track' }));

    expect(mockFn).toBeCalledTimes(1);
  });

  test('when the next button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [React.MouseEvent<HTMLButtonElement>]>();
    render(
      <MainControls
        isPlayed
        onClickNext={mockFn}
        showNextAndPreviousControls
      />
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button', { name: 'next track' }));

    expect(mockFn).toBeCalledTimes(1);
  });

  test('when the isPlayed prop is passed, the text in the button changes to pause', () => {
    render(
      <MainControls isPlayed />
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toHaveTextContent(/pause/i);
  });

  test('snapshot with default prop values', () => {
    render(
      <MainControls
        isPlayed
      />
    );

    expect(screen.getByTestId<HTMLDivElement>('main-controls')).toMatchSnapshot();
  });

  test('snapshot with showNextAndPreviousControls prop', () => {
    render(
      <MainControls
        isPlayed
        showNextAndPreviousControls
      />
    );

    expect(screen.getByTestId<HTMLDivElement>('main-controls')).toMatchSnapshot();
  });

  test('snapshot with isPlayed prop', () => {
    render(
      <MainControls
        isPlayed
      />
    );

    expect(screen.getByTestId<HTMLDivElement>('main-controls')).toMatchSnapshot();
  });
});
