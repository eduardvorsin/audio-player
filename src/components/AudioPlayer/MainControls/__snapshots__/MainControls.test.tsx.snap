// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume Controls component tests snapshot with default prop values 1`] = `
<div
  class="sc-bcXHqe bTCwa-d"
  data-testid="main-controls"
>
  <button
    class="sc-gswNZR gicUoi"
    data-testid="button"
  >
    <svg />
    play
  </button>
</div>
`;

exports[`Volume Controls component tests snapshot with isPlayed prop 1`] = `
<div
  class="sc-bcXHqe bTCwa-d"
  data-testid="main-controls"
>
  <button
    class="sc-gswNZR gicUoi"
    data-testid="button"
  >
    <svg />
    pause
  </button>
</div>
`;

exports[`Volume Controls component tests snapshot with showNextAndPreviousControls prop 1`] = `
<div
  class="sc-bcXHqe lkmtZM"
  data-testid="main-controls"
>
  <button
    class="sc-gswNZR gicUoi"
    data-testid="button"
  >
    <svg />
    prev track
  </button>
  <button
    class="sc-gswNZR gicUoi"
    data-testid="button"
  >
    <svg />
    play
  </button>
  <button
    class="sc-gswNZR gicUoi"
    data-testid="button"
  >
    <svg />
    next track
  </button>
</div>
`;
